src/app/games/hive/HivePiece.ts

```diff
@@ -0,0 +1,348 @@
+import { HiveMoveCoordToCoord, HiveMoveSpider } from './HiveMove';
+import { HiveState } from './HiveState';
+
+type HivePieceKind = 'QueenBee' | 'Beetle' | 'Grasshopper' | 'Spider' | 'SoldierAnt'
+
+export abstract class HivePiece implements ComparableObject {
+
+    public static encoder: Encoder<HivePiece> = new class extends Encoder<HivePiece> {
+
+        public encode(piece: HivePiece): JSONValue {
+            return { owner: piece.owner.value, kind: piece.kind };
+        }
+
+        public decode(encoded: JSONValue): HivePiece {
+            // eslint-disable-next-line dot-notation
+            assert(Utils.getNonNullable(encoded)['kind'] !== null, 'invalid encoded HivePiece');
+            // eslint-disable-next-line dot-notation
+            assert(Utils.getNonNullable(encoded)['owner'] !== null, 'invalid encoded HivePiece');
+            // eslint-disable-next-line dot-notation
+            const kind: string = Utils.getNonNullable(encoded)['kind'] as HivePieceKind;
+            // eslint-disable-next-line dot-notation
+            const owner: Player = Player.of(Utils.getNonNullable(encoded)['owner']);
+            switch (kind) {
+                case 'QueenBee': return new HivePieceQueenBee(owner);
+                case 'Beetle': return new HivePieceBeetle(owner);
+                case 'Grasshopper': return new HivePieceGrasshopper(owner);
+                case 'Spider': return new HivePieceSpider(owner);
+                default:
+                    Utils.expectToBe(kind, 'SoldierAnt');
+                    return new HivePieceSoldierAnt(owner);
+            }
+        }
+    };
+
+    protected constructor(public readonly owner: Player,
+                          public readonly kind: HivePieceKind,
+                          public readonly mustHaveFreedomToMove: boolean)
+    {
+    }
+
+    public toString(): string {
+        return `${this.kind}_${this.owner.toString()}`;
+    }
+
+    public equals(other: HivePiece): boolean {
+        return this.owner === other.owner && this.kind === other.kind;
+    }
+
+    protected destinationIsEmpty(move: HiveMoveCoordToCoord, state: HiveState): boolean {
+        return state.getAt(move.end).isEmpty();
+    }
+
+    protected canSlide(state: HiveState, start: Coord, end: Coord): boolean {
+        // For the piece to slide from start to end,
+        // one of the two common neighbors of start and coord should be empty
+        const startNeighbors: MGPSet<Coord> = new MGPSet(HexagonalUtils.neighbors(start));
+        const endNeighbors: MGPSet<Coord> = new MGPSet(HexagonalUtils.neighbors(end));
+        const commonNeighbors: MGPSet<Coord> = startNeighbors.intersect(endNeighbors);
+        for (const neighbor of commonNeighbors) {
+            if (state.getAt(neighbor).isEmpty()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public abstract moveValidity(move: HiveMoveCoordToCoord, state: HiveState): MGPFallible<void>
```

Comment:
Hélà monsieur l'pythoniste ! T'es point Virgile !

Reply:
errr c'est pas dégueu les points virgules là ? Genre ça fait comme si t'en mettais un après le `}` final d'une méthode.
Le `;` sépare les statements, ici c'est pas un statement ?

Reply:
c'est quoi un statement ?

Reply:
"In computer programming, a statement is a syntactic unit of an imperative programming language that expresses some action to be carried out"
Ici point d'action mon bon monsieur !
Mais allez je rajoute un ';', faudra trouver comment enforcer ça car je vais en oublier un bon paquet !